toss/fitness/fitness_function_utils.py

```diff
@@ -45,4 +45,68 @@ def _compute_squared_distance(positions,constant):
     """
-    return np.sum(np.power(positions,2), axis=0) - constant**2
\ No newline at end of file
+    return np.sum(np.power(positions,2), axis=0) - constant**2
+
+
+def compute_space_coverage(positions, velocities, timesteps, radius_outer_bounding_sphere):
+    """
+    In this function, we create a spherical meshgrid by defining a number of points
+    inside the outer bounding sphere. We then generate a multidimensional array
+    of boolean values corresponding to each point in the meshgrid. These values 
+    are initialy set to false since they have not been visited by any spacecraft.
+
+    Given a set of posiions on the trajectory that are defined inside the 
+    outer bounding sphere, we identify the points that are the closest the given positions, 
+    and subsequently reassign the visited boolean value to true. The ratio of visited points 
+    is then given by the number of True values to the total number of values in the boolean array.
+
+    Args:
+        positions (np.ndarray): (3,N) Array of positions (in cartesian coordinates).
+        velocities (np.ndarray): (3,N) Array of positions (in cartesian frame).
+        timesteps (np.ndarray): (N) Array of time values for each position.
+        radius_outer_bounding_sphere (float): Radius of outer bounding sphere.
+
+    Returns:
+        ratio (float): Number of True values to the total number of values in the boolean array
+    """
+
+    # Define frequency of points for the spherical meshgrid: (see: Courant–Friedrichs–Lewy condition)
+    cfl = 1 # Courant number
+    v_max = np.sqrt(np.max(velocities[0,:]**2 + velocities[1,:]**2 + velocities[2,:]**2))
+    dt = timesteps[1]-timesteps[0]
+    
+    # Calculate and adjust grid spacing based on the Courant number, maximal velocity, and time step
+    grid_spacing = v_max * dt / cfl
+    r = np.linspace(0, radius_outer_bounding_sphere, int(radius_outer_bounding_sphere / grid_spacing)) # Number of evenly spaced points along the radial axis
+    theta = np.linspace(0, np.pi, int(np.pi / grid_spacing)) # Number of evenly spaced points along the polar angle (defined in between 0 and pi)
+    phi = np.linspace(0, 2 * np.pi, int(2 * np.pi / grid_spacing)) # Number of evenly spaced points along the azimuthal angle (defined in between 0 and 2*pi)
+
+    # Create a spherical meshgrid
+    r_matrix, theta_matrix, phi_matrix = np.meshgrid(r, theta, phi)
+
+    # Convert the positions along the trajectory to spherical coordinates
+    r_points = np.sqrt(positions[0,:]**2 + positions[1,:]**2 + positions[2,:]**2) # radial coordinates of points
+    theta_points = np.arccos(positions[2,:] / r_points) # polar angle coordinates of points
+    phi_points = np.arctan2(positions[1,:], positions[0,:]) # azimuthal angle coordinates of points
+
+    # Remove points outside measurement zone (i.e outside outer-bounding sphere)
+    index_feasible_positions = r_points < radius_outer_bounding_sphere
```

Comment:
shouldn't you also have to remove the ones with too small radii? :thinking: what happens for collisions?